packages/apps/plugins/plugin-thread/src/components/MessageContainer.tsx

```diff
@@ -0,0 +1,133 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { DotsSixVertical, X } from '@phosphor-icons/react';
+import React, { forwardRef, useState } from 'react';
+
+import { type Message as MessageType } from '@braneframe/types';
+import { Surface } from '@dxos/app-framework';
+import { type SpaceMember } from '@dxos/client/echo';
+import { type Expando, getTextContent, TextObject } from '@dxos/echo-schema';
+import { PublicKey } from '@dxos/react-client';
+import { Button } from '@dxos/react-ui';
+import { TextEditor, useTextModel } from '@dxos/react-ui-editor';
+import { Mosaic, type MosaicTileComponent } from '@dxos/react-ui-mosaic';
+import { hoverableControlItem, hoverableControls, hoverableFocusedWithinControls, mx } from '@dxos/react-ui-theme';
+import { Message, type MessageBlockProps, type MessageProps } from '@dxos/react-ui-thread';
+
+import { useMessageMetadata } from '../hooks';
+import { THREAD_ITEM } from '../meta';
+
+type Block = MessageType.Block;
+
+const safeParseJson = (data: string) => {
```

Comment:
TODO to make util (this isn't the first time https://github.com/dxos/dxos/blob/main/packages/apps/composer-app/src/native.ts#L88)